### PR TITLE
Convert Lit-GPT Falcon Format to Original Falcon Format

### PR DESCRIPTION
**Pull request by JustinGoheen**
This PR converts a Lit-GPT model back to Huggingface format by reversing the operations of `convert_hf_checkpoint.py` as stated in this comment: https://github.com/Lightning-AI/lit-gpt/issues/183#issuecomment-1618843317

### CODE REVIEW

**Comment by JustinGoheen** [comment]
I downloaded and tested the following models: OpenLLaMA, Falcon-7B, and Pythia-1B.

The tests are written in a manner that requires:
1. downloaded each model
2. converting each model to Lit-GPT format
3. converting the Lit-GPT formatted model back to HF format

The tests pass locally
Screenshot 2023-07-19 at 2 49 32 PM

> **Note**
> The 1 skipped test is for Falcon 40b

**Comment by JustinGoheen** [comment]
moved back to draft because I added an assertion to check the length of the converted keys matches the length of the original keys, and tests fail on those assertions for missing keys in `copy_weights_gpt_neox` and `copy_weights_hf_llama`.

**Comment by lantiga** [comment]
Thank you @JustinGoheen!
Just calling out that tests don’t need the model to be downloaded, we can initialize with random weights and test the roundtrip consistency.

File: tests/test_convert_lit_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import torch
+import json
+import os
+import pytest
+
+
+@pytest.mark.parametrize("model_size", ["7b", "40b"])
+def test_copy_weights_falcon(model_size):
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:8]: In https://github.com/Lightning-AI/lit-gpt/blob/main/tests/test_model.py we test our implementation by

1. Loading the HF model: https://github.com/Lightning-AI/lit-gpt/blob/main/tests/test_model.py#L105-L109
2. Copying over its weights to our model: https://github.com/Lightning-AI/lit-gpt/blob/main/tests/test_model.py#L110-L114
3. Running a sample through the model and comparing the result: https://github.com/Lightning-AI/lit-gpt/blob/main/tests/test_model.py#L116-L120

I suggest that you do exactly the same, but copying our weights to the HF model.

This means you won't need to download checkpoints or any sort of configuration file

Another interesting way to test that conversion works would be to do

```python
our_state_dict = our_model.state_dict()
to_hf = convert_lit_checkpoint(our_state_dict)
to_ours = convert_hf_checkpoint(to_hf)
assert to_ours == our_state_dict
```

File: scripts/convert_lit_checkpoint.py [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    split[idx] = "{}"
+    from_name = ".".join(split)
+    return from_name, number
+
+
+def load_param(param: Union[torch.Tensor, NotYetLoadedTensor]) -> torch.Tensor:
+    if hasattr(param, "_load_tensor"):
+        # support tensors loaded via `lazy_load()`
+        return param._load_tensor()
+    return param
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:180]: I would import these from the other script if it's the same

**Review by carmocca — COMMENTED** [review]
You don't need to support conversion for all models together in one PR. I would start with the gpt neox model which is probably the easiest

**Review by JustinGoheen — COMMENTED** [review]
need to concat tensors correctly

**Comment by JustinGoheen** [comment]
The conversion and test are ready for Falcon. The conversion is ready for LLaMA and GPT-NeoX however, tests are not ready. I'd like to have this PR merged as a conversion for Falcon, and then work on the tests for LLaMA and GPT-NeoX in separate PRs.

**Comment by lantiga** [comment]
> The conversion and test are ready for Falcon. The conversion is ready for LLaMA and GPT-NeoX however, tests are not ready. I'd like to have this PR merged as a conversion for Falcon, and then work on the tests for LLaMA and GPT-NeoX in separate PRs.

That's fine with me but let's not drop the ball and follow up with tests for the other models asap.

Also we need to clarify that this is only supported for Falcon in the instructions, and break with an informative message if you pass different checkpoints.

**Comment by JustinGoheen** [comment]
> > The conversion and test are ready for Falcon. The conversion is ready for LLaMA and GPT-NeoX however, tests are not ready. I'd like to have this PR merged as a conversion for Falcon, and then work on the tests for LLaMA and GPT-NeoX in separate PRs.
> 
> That's fine with me but let's not drop the ball and follow up with tests for the other models asap.
> 
> Also we need to clarify that this is only supported for Falcon in the instructions, and break with an informative message if you pass different checkpoints.

I updated `convert_lit_checkpoint` to raise a NotImplementedError for models other than Falcon. I'll also submit issues for the other tests – please assign those to me.

File: tests/test_convert_lit_checkpoint.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+@pytest.mark.skip("not implemented")
+@pytest.mark.parametrize("size", ("7b", "70b"))
+def test_against_hf_llama2(size):
+    pass
+
+
+@torch.inference_mode()
+@pytest.mark.skip("not implemented")
+def test_against_gpt_neox(size):
+    pass
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:78]: These stubs shouldn't be added, they can be added directly in their corresponding PR's once they are ready

File: scripts/convert_lit_checkpoint.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+# support running without installing as a package
+wd = Path(__file__).parent.parent.resolve()
+sys.path.append(str(wd))
+
+from lit_gpt import Config
+from lit_gpt.utils import lazy_load, incremental_save, NotYetLoadedTensor
+from scripts.convert_hf_checkpoint import load_param, layer_template
+
+
+def copy_weights_gpt_neox(
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:19]: The untested functions shouldnt be added in this PR, instead add them with their corresponding PRs

File: scripts/convert_lit_checkpoint.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+    for k, v in weight_map.items():
+        if lit_key_name == v:
+            return k
+
+
+@torch.inference_mode()
+def convert_lit_checkpoint(
+    *,
+    checkpoint_dir: Path = Path("checkpoints/tiiuae/falcon-7b"),
+    checkpoint_name: str = "lit_model_finetuned.pth",
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:207]: Since the checkpoint name is different depending on the script and if it's not an end-of-training checkpoint this should be made a required argument (not optional) to avoid silently converting the wrong checkpoint

File: scripts/convert_lit_checkpoint.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    else:
+        raise NotImplementedError(f"Conversion for {model_name} is not yet supported")
+
+    # initialize a new empty state dict to hold our new weights
+    sd = {}
+
+    # checkpoint_name cannot be hardcoded because there exists different outputs such as
+    # ("lit_model_finetuned.pth", "lit_model_lora_finetuned.pth", "lit_model_adapter_finetuned.pth"")
+    pth_file = checkpoint_dir / checkpoint_name
+    bin_file = checkpoint_name.replace(".pth", ".bin")
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:225]: Why is this replacement necessary?

File: tests/test_convert_lit_checkpoint.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+def test_against_original_falcon_40b():
+    file_path = wd / "tests" / "original_falcon_40b.py"
+    url = "https://gist.githubusercontent.com/carmocca/feed39b1bc65a29f73c1cecc58a01167/raw/a9a65f2b93716b3c09ec9f354d535ae5953de08f/original_falcon_40b.py"
+    if not file_path.is_file():
+        urlretrieve(url=url, filename=file_path)
+
+    from tests.original_falcon_40b import RWConfig, RWForCausalLM
+    from lit_gpt import Config, GPT
+    from scripts.convert_lit_checkpoint import copy_weights_falcon
+    from scripts.convert_hf_checkpoint import copy_weights_falcon as copy_weights_falcon_hf
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:20]: This would make the code easier to read IMO

```suggestion
    from scripts.convert_lit_checkpoint import copy_weights_falcon as copy_to_theirs
    from scripts.convert_hf_checkpoint import copy_weights_falcon as copy_to_ours
```

File: tests/test_convert_lit_checkpoint.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+    # the unchanged model config
+    our_config = Config.from_name("falcon-40b", n_layer=2, n_head=8, n_query_groups=4, n_embd=32)
+    theirs_config = RWConfig(
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:24]: "ours" or "theirs" is meant to answer, "whose?"

So if I'm not mistaken, this should be "ours" instead of "our" to be consistent with "theirs" and the other script

File: tests/test_convert_lit_checkpoint.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        hidden_size=32, n_head=8, n_head_kv=4, n_layer=2, parallel_attn=True, vocab_size=65024, bias=False
+    )
+
+    # need to begin with weight initialized from a ground truth source
+    ground_truth_their_model = RWForCausalLM(theirs_config)
+    ground_truth_their_state_dict = ground_truth_their_model.state_dict()
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:30]: Why is this necessary? Conversion should work if you start with our model.

File: tests/test_convert_lit_checkpoint.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    # else a RuntimeError would occur while loading
+    their_model.load_state_dict(their_state_dict)
+    # check that all tensors that went into the conversion
+    # are saved in the correct order
+    assert all(
+        [
+            torch.equal(their_state_dict[tsdk], our_state_dict[osdk])
+            for tsdk, osdk in zip(their_state_dict.keys(), our_state_dict.keys())
+        ]
+    )
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:51]: This check is irrelevant if the assertions below already pass

File: scripts/convert_lit_checkpoint.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+        )
+    else:
+        raise NotImplementedError
+
+    for name, param in lit_weights.items():
+        if "transformer.h" in name:
+            from_name, number = layer_template(name, 2)
+            to_name = get_to_name(from_name, weight_map).format(number)
+        else:
+            to_name = get_to_name(name, weight_map)
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py:102]: I did that originally but then changed it back because of the None keys. If we have to return a state_dict with all weight names, I think we would need to keep this as is.

File: tests/test_convert_lit_checkpoint.py (hunk 10 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: removed in https://github.com/Lightning-AI/lit-gpt/pull/273/commits/ee894af7f93c27691fbcbfbe5da3b1349355fe5a

File: tests/test_convert_lit_checkpoint.py (hunk 9 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: improved in https://github.com/Lightning-AI/lit-gpt/pull/273/commits/ee894af7f93c27691fbcbfbe5da3b1349355fe5a

File: scripts/convert_lit_checkpoint.py (hunk 6 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py:225]: I noticed all HF models are bin and not pth; so I converted just in case HF requires .bin file ext. If they do not, then the replacement is not needed. 

The replacement, though not really necessary, could be useful to distinguish model formats in addition to `lit_model` for our format.

File: tests/test_convert_lit_checkpoint.py (hunk 3 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: removed

File: tests/test_convert_lit_checkpoint.py (hunk 8 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: updated in https://github.com/Lightning-AI/lit-gpt/pull/273/commits/ee894af7f93c27691fbcbfbe5da3b1349355fe5a

Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: I changed this to `config` only because we need to use the same model, so there is only a need for 1 config

File: scripts/convert_lit_checkpoint.py (hunk 5 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: implemented in [c874fa2](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/c874fa23187a324f6161283de2020786c4fd0d11)

File: .gitignore [hunk 12]
```diff
@@ -3,6 +3,10 @@ __pycache__
 .DS_Store
 *.egg-info
 build
+.venv
+.vscode
+*ruff*
+pyproject.toml
```
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply .gitignore:9]: these can be discarded if we want a minimal .gitignore

File: scripts/convert_lit_checkpoint.py (hunk 4 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: removed in [4becaa5](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/4becaa52afc748ca1572ea5b38e9439bb8c2781b)

File: scripts/convert_lit_checkpoint.py (hunk 6 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: changed the convention in [16c9675](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/16c96750923a820b384c63bed672c49aa60e6b50)

users will now be prompted in the tutorial to pass the checkpoint_name without a file extension, and .bin will be joined to the given checkpoint_name

File: .gitignore (hunk 12 above)
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply .gitignore]: removed in [8fc4835](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/8fc48351e35af58c8ff4d0661a1761229d9bbf84)

File: scripts/convert_lit_checkpoint.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py:60]: I can invert for Falcon since there are no None keys; however, I made need to add this back in for GPT-X and Llama-2.

[the remark this answers is not in the document]
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: implemented in [f24ec91](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/f24ec91770c6fecf058a6f9361aef36e853f31fe)

File: scripts/convert_lit_checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+            to_name = weight_map[from_name].format(number)
+        else:
+            to_name = weight_map[name]
+        param = load_param(param)
+        if saver is not None:
+            param = saver.store_early(param)
+        state_dict[to_name] = param
+
+
+def get_to_name(lit_key_name: str, weight_map: Dict[str, str]) -> str:
```
Comment (carmocca) [inline scripts/convert_lit_checkpoint.py:67]: This is now unused

File: tests/test_convert_lit_checkpoint.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+import pytest
+import torch
+
+wd = Path(__file__).parent.parent.absolute()
+
+
+def test_convert_lit_checkpoint(tmp_path):
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    chkpt_name = "lit_model_finetuned"
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:14]: nit: usually we write ckpt

```suggestion
    ckpt_name = "lit_model_finetuned"
```

File: tests/test_convert_lit_checkpoint.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    )
+
+    their_model = RWForCausalLM(config)
+    their_state_dict = their_model.state_dict()
+
+    ours_state_dict = {}
+    copy_to_ours("40b", ours_state_dict, their_state_dict)
+
+    converted_state_dict = {}
+    copy_to_theirs("40b", converted_state_dict, ours_state_dict)
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:50]: Shouldn't this work if we started off our model directly?

```suggestion
    our_model = GPT(our_config)
    ours_state_dict = our_model.state_dict()

    converted_state_dict = {}
    copy_to_theirs("40b", converted_state_dict, ours_state_dict)
```

If not, why?

File: tutorials/convert_lit_models.md [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+> each example shown below uses Falcon-7B <br/>
+> please be sure to update --checkpoint_dir accordingly
+
+### Full Finetuning
+
+After full finetuning, your checkpoint directory will contain a file named `lit_model_finetuned.pth`and converting the finetuned model back to its original weights naming convention can be done by setting the `--checkpoint_name` with:
+
+```sh
+python scripts/convert_lit_checkpoint.py \
+    --checkpoint_dir=checkpoints/tiiuae/falcon-7b
```
Comment (carmocca) [inline tutorials/convert_lit_models.md:18]:
```suggestion
    --checkpoint_dir=checkpoints/tiiuae/falcon-7b \
```

File: tutorials/convert_lit_models.md [hunk 17]
```diff
@@ -0,0 +1,40 @@
+## Converting a Lit-GPT Formatted Model to its Original Format
+
+Several models retrieved from external sources must be reformatted with naming conventions for Lit-GPT weights before finetuning. We've provided a helpful script to convert models finetuned with Lit-GPT back to their original format.
+
+The way the script is used depends on the finetuning method. The commands for converting a model back to its original format after using one of the several finetuning methods supported by Lit-GPT are shown below.
+
+> **Note**
```
Comment (carmocca) [inline tutorials/convert_lit_models.md:7]: The "Note" notation doesn't work with line breaks unfortunately. If you look at the rendered markdown, you'll see it appears as a "quote"

**Review by carmocca — COMMENTED** [review]
Looking great Justin! The reduced scope makes everything easier. Just on last main question about how you are testing

File: tutorials/convert_lit_models.md (hunk 17 above)
Comment (JustinGoheen) [reply tutorials/convert_lit_models.md:7]: seems this admonition style broke in the last 24 hours: https://github.com/orgs/community/discussions/16925

Comment (JustinGoheen) [reply tutorials/convert_lit_models.md:7]: I commited a change to remove the space and line break - which works here, but not in the rendered .md file

> **Note**
> each example shown below uses Falcon-7B please be sure to update --checkpoint_dir accordingly

Comment (JustinGoheen) [reply tutorials/convert_lit_models.md:7]: the new syntax for .md file is

```md
> [!NOTE]  
> ipsum lorem
```

> [!NOTE]  
> ipsum lorem

File: scripts/convert_lit_checkpoint.py (hunk 13 above)
Comment (JustinGoheen) [reply scripts/convert_lit_checkpoint.py]: removed in https://github.com/Lightning-AI/lit-gpt/pull/273/commits/6af2d81729dbca03404ee4230975e3e5e8a2a610

File: tests/test_convert_lit_checkpoint.py (hunk 15 above)
Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py:50]: when I do:

```python
    ours_model = GPT(ours_config)
    ours_state_dict = ours_model.state_dict()

    theirs_state_dict = {}
    copy_to_theirs("40b", theirs_state_dict, ours_state_dict)

    theirs_model = RWForCausalLM(theirs_config)
    theirs_model.load_state_dict(theirs_state_dict)

    # test end to end
    x = torch.tensor([[9856, 23, 491, 1536, 304]], dtype=torch.int32)
    theirs_y = theirs_model(x)["logits"]
    ours_y = ours_model(x)
    torch.testing.assert_close(theirs_y, ours_y)
```

I get:

`FAILED tests/test_convert_lit_checkpoint.py::test_against_original_falcon_40b - AssertionError: Tensor-likes are not close!`

with:

```
>       torch.testing.assert_close(converted_y, ours_y)
E       AssertionError: Tensor-likes are not close!
E       
E       Mismatched elements: 325117 / 325120 (100.0%)
E       Greatest absolute difference: 3.43747615814209 at index (0, 0, 58720) (up to 1e-05 allowed)
E       Greatest relative difference: 190444.875 at index (0, 1, 64161) (up to 1.3e-06 allowed)
```

Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py:50]: I had done something similar to this at first, and this error/failure is why I began writing the assertions to check the length of the state_dicts, the returned param tensors, and that the key names in the converted state_dict matched a ground truth theirs_model.state_dict ... all of those assertions passed, however - the torch.testing.assert_close does not.

Comment (JustinGoheen) [reply tests/test_convert_lit_checkpoint.py]: @carmocca @lantiga this is resolved as of [3e6b412](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/3e6b41208c7385101e84743eedb1b1644821a139).

The fix was setting `assign` to True in `theirs.load_state_dict()`.

File: tests/test_convert_lit_checkpoint.py [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+        bias=False,
+    )
+
+    ours_model = GPT(ours_config)
+    ours_state_dict = ours_model.state_dict()
+    theirs_state_dict = {}
+    copy_to_theirs("40b", theirs_state_dict, ours_state_dict)
+
+    theirs_model = RWForCausalLM(theirs_config)
+    # assign must be set to True for torch.testing.assert_close to pass
```
Comment (carmocca) [inline tests/test_convert_lit_checkpoint.py:56]: TIL about `assign=True`. Seems to be a recent addition https://github.com/pytorch/pytorch/pull/102212. cc @awaelchli for visibility

File: tutorials/convert_lit_models.md (hunk 17 above)
Comment (carmocca) [reply tutorials/convert_lit_models.md]: @JustinGoheen Could you fix the rest of the usages in this repo in a separate PR?

**Review by carmocca — APPROVED** [review]
Let's go!

File: tutorials/convert_lit_models.md (hunk 17 above)
Comment (JustinGoheen) [reply tutorials/convert_lit_models.md]: will do. see #305
